### PR TITLE
feat: support more advanced suite of LIKE expressions

### LINKED ISSUE #2054: Support more general use of '%' and '_' wildcards in LIKE expressions
'%' means match zero or more characters
'_' means match a single character

Currently, we only support '%', and only at the beginning or end of the expression.

We should support '_', and it should be possible to have either wildcard within the LIKE expression.

### PR DESCRIPTION
**Pull request by agavra**
fixes #2054 

### Description 

Adds support for more advanced LIKE functionality by turning a LIKE predicate into a regex statement and evaluating that. I looked for existing java implementations but none seemed much better than implementing it myself and avoiding a third-party (heavyweight) dependency.

This change also allows you to reference a column now in the LIKE predicate.

### Testing done 

- Extensive Unit test
- QTT test

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/helpers/LikeEvaluator.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  /**
+   * @param val       the value to match
+   * @param pattern   the pattern to match against
+   *
+   * @return  whether or not {@code val} matches {@code pattern} given the SQL
+   *          LIKE definition of matching
+   */
+  public static boolean matches(final String val, final String pattern) {
+    // note that we do not yet support escape characters in the pattern
+    // see issue
```
Comment (big-andy-coates) [inline ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/helpers/LikeEvaluator.java:48]: missing issue link?

File: ksqldb-execution/src/test/java/io/confluent/ksql/execution/codegen/helpers/LikeEvaluatorTest.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    // name | val | pattern
+    return new Object[][]{
+        {"nothing special",               "foo", "foo"},
+        {"nothing special [X]",           "bar", "foo"},
+        {"empty percents",                "foo", "%foo%"},
+        {"empty percents [X]",            "bar", "%foo%"},
+        {"percents",                      "barfoobaz", "%foo%"},
+        {"percents [X]",                  "barbarbar", "%foo%"},
+        {"percents one side",             "barfoo", "%foo"},
+        {"percents one side [X]",         "barbarbar", "%foo"},
```
Comment (big-andy-coates) [inline ksqldb-execution/src/test/java/io/confluent/ksql/execution/codegen/helpers/LikeEvaluatorTest.java:45]:
```suggestion
        {"percents one side [X]",         "barfoobar", "%foo"},
```

Comment (big-andy-coates) [reply ksqldb-execution/src/test/java/io/confluent/ksql/execution/codegen/helpers/LikeEvaluatorTest.java:45]: also, worth adding case for:
*  % other side and 
* % in middle and 
* multiple %s
* escaped %s???? Or do we not support them yet.. (hard to add?)

**Comment by agavra** [comment]
> Thanks @agavra
> 
> LGTM, 'cept where does this character range stuff come from? I've not seen that in SQL before. Is that in the spec? AFAIK SQL LIKE only has `_` and `%` with special meaning...
> 
> (I'm only putting a blocker on as I think we should think carefully about non-SQL syntax, if indeed it is).

Huh - I looked at https://docs.microsoft.com/en-us/sql/t-sql/language-elements/like-transact-sql?view=sql-server-ver15 since they tend to have the best docs and just assumed that it was standard. But looking at postgres they don't follow the same mechanism.... I'm happy to go either way about it, but that was the hardest to code so I'd low-key be bummed about removing it 😂

File: ksqldb-execution/src/test/java/io/confluent/ksql/execution/codegen/helpers/LikeEvaluatorTest.java (hunk 2 above)
Comment (agavra) [reply ksqldb-execution/src/test/java/io/confluent/ksql/execution/codegen/helpers/LikeEvaluatorTest.java:45]: the middle two are covered by `c[a-c]t%[a-e][^abc]%m__w` - i'll add tests for the first and last (escaping is done by `[%]` in MSFT-sql at least and should be supported here as well)

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/helpers/LikeEvaluator.java (hunk 1 above)
Comment (agavra) [reply ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/helpers/LikeEvaluator.java:48]: https://github.com/confluentinc/ksql/issues/5021 will add it

**Comment by blueedgenick** [comment]
@agavra can i chime in with another "looks good but it's in the wrong function" observation here ? I definitely like getting more search functionality but that SQLServer implementation is, unfortunately, not a very representative one to model from. The more"usual" implementation is to only support '_' and '%' in the `LIKE` function, perhaps with an optional escape character argument, and then have a `REGEXP_LIKE` to handle the fancier stuff. See examples here (https://prestodb.io/docs/current/functions/regexp.html) and (https://docs.oracle.com/en/database/oracle/oracle-database/20/sqlrf/Pattern-matching-Conditions.html#GUID-0779657B-06A8-441F-90C5-044B47862A0A)

**Comment by agavra** [comment]
Well it seems I've learned my lesson - I will check docs other than MSFT SqlServer before implementing difficult functionality 😂 I'll remove the `[]` stuff

**Comment by big-andy-coates** [comment]
Can you _add_ the escape char too?

**Comment by agavra** [comment]
> Can you _add_ the escape char too?

If you give a mouse a cookie 🍪 .... haha sure I'll add it too

**Comment by agavra** [comment]
Something very weird is happening with git - please hold on reviewing this until I fix it

File: ksqldb-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java [hunk 3]
```diff
@@ -889,10 +889,21 @@ public Node visitNullPredicate(final SqlBaseParser.NullPredicateContext context)
 
     @Override
     public Node visitLike(final SqlBaseParser.LikeContext context) {
+
+      final Optional<String> escape = Optional.ofNullable(context.escape)
+          .map(Token::getText)
+          .map(s -> ParserUtil.unquote(s, "'"));
+      escape.ifPresent(s -> {
+        if (s.length() != 1) {
+          throw new IllegalStateException("Expected single character escape but got: " + s);
```
Comment (big-andy-coates) [inline ksqldb-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java:898]: IllegalStateException?

The syntax allows the escape to be a multi-char string, so this is the error the users going to see if they enter one, right?

I'd of thought this should be :
```suggestion
          throw new KsqlException(loc.asPrefix() + "Expected single character escape but got: " + s);
```

Where `loc` is the `NodeLocation`.

Comment (agavra) [reply ksqldb-parser/src/main/java/io/confluent/ksql/parser/AstBuilder.java:898]: ah good catch, originally I had it in the syntax as just a single char (`.`) but then i changed it to `STRING` to include the quotes and forgot to change this!

File: ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/SqlToJavaVisitor.java [hunk 4]
```diff
@@ -724,46 +726,23 @@ private String buildSupplierCode(final String typeString, final String code) {
...
-        } else {
-          return new Pair<>(
-              "(" + valueString + ").endsWith(\"" + patternString.substring(1) + "\")",
-              SqlTypes.STRING
-          );
-        }
-      }
 
-      if (patternString.endsWith("%")) {
+      if (node.getEscape().isPresent()) {
```
[the remark this answers is not in the document]
Comment (agavra) [reply ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/SqlToJavaVisitor.java:732]: I will expose the private method for testing

[the remark this answers is not in the document]
Comment (agavra) [reply ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/SqlToJavaVisitor.java:732]: somewhat - I'd still need to do:
```java
LikeEvaluator.matches(... + "Optional.ofNullable(" + escape.map(escape -> "'" + escape + "'").getOrElse(null))
```
And to be honest I don't think that's any cleaner

**Pull request merged**